frontend/packages/console-dynamic-plugin-sdk/src/webpack/ConsoleRemotePlugin.ts

```diff
@@ -35,22 +38,98 @@ export const validatePackageFileSchema = (
   return validator.result;
 };
 
+const validateConsoleProvidedSharedModules = (
+  pkg: ConsolePackageJSON,
+  sdkPkg = loadPackageManifest('@openshift-console/dynamic-plugin-sdk'),
+) => {
+  const pluginDeps = { ...pkg.devDependencies, ...pkg.dependencies };
+  const result = new ValidationResult('package.json');
+
+  sharedPluginModules.forEach((moduleName) => {
+    // Skip modules for which a fallback (different) version can be provided by the plugin.
+    // Also skip modules which are not explicitly listed in the plugin's dependencies.
+    if (sharedPluginModulesMetadata[moduleName].allowFallback || !pluginDeps[moduleName]) {
+      return;
+    }
+
+    const providedVersionRange = sdkPkg.dependencies[moduleName];
+    const consumedVersion = loadPackageManifest(moduleName).version;
+
+    if (semver.validRange(providedVersionRange) && semver.valid(consumedVersion)) {
+      result.assertThat(
+        semver.satisfies(consumedVersion, providedVersionRange),
+        `Console provides shared module ${moduleName} ${providedVersionRange} but plugin uses version ${consumedVersion}`,
+      );
+    }
+  });
+
+  return result;
+};
+
+export type ConsoleRemotePluginOptions = Partial<{
+  /**
+   * Validate the plugin's `package.json` file schema?
```

Comment:
I would add some more detailed comments about what each of these means. (Presumably this shows up in the hover in VSCode?)

Let's add this to the README as well.